NIFI-8650 Flow Analysis

tpalfy opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-8650](https://issues.apache.org/jira/browse/NIFI-8650)

Created new component: FlowAnalyisRule.
Added new framework functionality: analyze flow.

FlowAnalysisRules can be managed from the UI: hamburger-menu -> Controller Settings -> Flow Analysis Rules

FlowAnalysisRules can be set up with an enforcement policy (on the SETTINGS tab) that can have one of 2 values: WARN - Creates warnings through logs and bulletins. Doesn't impact the flow otherwise. ENFORCE - Apart from the warnings, also invalidates components that violate the rule.

Flow analysis is run by the framework:
-For the whole flow (starting at the root process group), scheduled every 5 minutes (similar to validation) -For each component individually, during their validation -For a process group (can be root), when requested asynchronously through REST interface (not from UI as of yet)

Flow analysis may produce rule violations as a result. These are stored in memory and are not persisted. They
-Can be requested through REST interface (not from UI as of yet) -In case of rules with ENFORCE enforcement policy, invalidate the corresponding components

Currently, FlowAnalysisRule implementations are directly provided either -a component (currently supported Processors and ControllerServices) as VersionedComponent instances. Concrete type should be determined by the rule. -a process group as a VersionedProcessGroup.
Indirectly they can analyze other components accessible from the directly provided components.

Added one FlowAnalyisRule implementation: DisallowComponentType. It can be configured with the (simple or fully qualified) name of a type. Produces a rule violation for each supported component (currently controller service and processor) instances of that type.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

sardell commented:
@tpalfy I tried building the latest push but it's failing due to a compilation error:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:10 min (Wall Clock)
[INFO] Finished at: 2023-04-27T15:31:31-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project nifi-framework-cluster: Compilation failure: Compilation failure:
[ERROR] /Users/sardell/projects/nifi/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/test/java/org/apache/nifi/cluster/manager/FlowAnalysisResultEntityMergerTest.java:[233,30] error: incompatible types: Set<FlowAnalysisRuleDTO> cannot be converted to List<FlowAnalysisRuleDTO>
[ERROR] /Users/sardell/projects/nifi/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/test/java/org/apache/nifi/cluster/manager/FlowAnalysisResultEntityMergerTest.java:[234,39] error: incompatible types: Set<FlowAnalysisRuleViolationDTO> cannot be converted to List<FlowAnalysisRuleViolationDTO>
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/processor/StandardValidationContextFactory.java:
```diff
@@ -20,25 +20,53 @@
+    public StandardValidationContextFactory(
+        final ControllerServiceProvider serviceProvider,
+        final VariableRegistry variableRegistry,
+        final RuleViolationsManager ruleViolationsManager,
+        final FlowAnalyzer flowAnalyzer
```
Should these properties be added to the ValidationContextFactory? Property Validation and Flow Analysis are different operations, so combining them does not seem like the right approach.

tpalfy replied:
I think it makes sense if we don't consider the concept of Validation as validation of the properties of a component but the validation of the component itself.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -776,6 +779,23 @@ protected Collection<ValidationResult> computeValidationErrors(final ValidationC
 
+            analyze();
+
+            Optional.ofNullable(getValidationContextFactory().getRuleViolationsManager())
+                .map(ruleViolationsManager -> ruleViolationsManager.getRuleViolationsForSubject(getIdentifier()))
+                .map(Collection::stream)
+                .ifPresent(ruleViolationStream -> ruleViolationStream
+                    .filter(ruleViolation -> ruleViolation.getEnforcementPolicy() == EnforcementPolicy.ENFORCE)
+                    .filter(RuleViolation::isEnabled)
+                    .forEach(ruleViolation -> validationResults.add(
+                        new ValidationResult.Builder()
+                            .subject(getComponent().getClass().getSimpleName())
+                            .valid(false)
+                            .explanation(ruleViolation.getViolationMessage())
+                            .build()
+                    )
+                ));
```
It seems like Flow Analysis results should be separate from property validation failures.

tpalfy replied:
Again, this makes sense if we consider Validation as the validation of the component itself. Enforcing rules have the duty to invalidate violating components. I see no better place than doing it during validation.

markap14 replied:
I think it does make sense to go ahead and perform the rule analysis here. Since that analysis may be enforced by invalidating the component, we need to ensure that the analysis is performed as part of the validation. If they are performed out of sync, it can lead to problems where a component is valid, started, and then the rule enforcement indicates a problem.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/flow/FlowManager.java:
```diff
@@ -403,10 +406,50 @@ ParameterContext createParameterContext(String id, String name, Map<String, Para
+        final BundleCoordinate bundleCoordinate
+    );
+
+    FlowAnalysisRuleNode createFlowAnalysisRule(
+        final String type,
+        final BundleCoordinate bundleCoordinate,
+        final boolean firstTimeAdded
+    );
+
+    FlowAnalysisRuleNode createFlowAnalysisRule(
+        final String type,
+        final String id,
+        final BundleCoordinate bundleCoordinate,
+        final boolean firstTimeAdded
+    );
+
+    FlowAnalysisRuleNode createFlowAnalysisRule(
+        final String type,
+        final String id,
+        final BundleCoordinate bundleCoordinate,
+        final Set<URL> additionalUrls,
+        final boolean firstTimeAdded,
+        final boolean register,
+        final String classloaderIsolationKey
+    );
```
Is it necessary to define all of these overloaded methods as opposed to just one or two?

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/flowanalysis/AnalyzeFlowRequest.java:
```diff
@@ -0,0 +1,84 @@
+public class AnalyzeFlowRequest implements AnalyzeFlowStatus {
+    private final String processGroupId;
+
+    private final long submissionTime = System.currentTimeMillis();
```
Recommend aligning the property name with the accessor name:
```suggestion
    private final long requestSubmissionTime = System.currentTimeMillis();
```

tpalfy replied:
This class is not needed any more and is going to be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/flowanalysis/FlowAnalysisRuleInstantiationException.java:
```diff
@@ -0,0 +1,31 @@
+
+public class FlowAnalysisRuleInstantiationException extends Exception {
+
+    private static final long serialVersionUID = 189234789237L;
```
Recommend using `1L` as a starting value instead of this random value.
```suggestion
    private static final long serialVersionUID = 1L;
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/validation/RuleViolation.java:
```diff
@@ -0,0 +1,209 @@
+    private final String violationMessage;
+    private final String violationExplanation;
+
+    private boolean enabled;
```
The `enabled` property seems unclear in this context. If a Flow Analysis Rule is disabled, it should not generate Rule Violations, correct?

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/pom.xml:
```diff
@@ -323,6 +323,18 @@
                     </excludes>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>test-jar</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
Is this addition required?

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/analyzeflow/ruleimpl/DisallowComponentType.java:
```diff
@@ -0,0 +1,84 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.analyzeflow.ruleimpl;
```
Package names should not include `impl`. As an extension component, this rule class should be moved to a separate NAR.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/analyzeflow/ruleimpl/DisallowComponentType.java:
```diff
@@ -0,0 +1,84 @@
+            if (encounteredComponentType.equals(componentType) || encounteredSimpleComponentType.equals(componentType)) {
+                ComponentAnalysisResult result = ComponentAnalysisResult.newResult(
+                        "default",
+                        "'" + componentType + "' is not allowed!"
```
Exclamation marks should not be used in messages.
```suggestion
                        "'" + componentType + "' is not allowed"
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/ExtensionBuilder.java:
```diff
@@ -432,10 +452,57 @@ public ControllerServiceNode buildControllerService() {
+        try {
+            loggableComponent = createLoggableFlowAnalysisRule();
+        } catch (final FlowAnalysisRuleInstantiationException rtie) {
+            logger.error("Could not create FlowAnalysisRule of type " + type + " for ID " + identifier + "; creating \"Ghost\" implementation", rtie);
```
Placeholders should be used instead of string concatenation.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java:
```diff
@@ -1156,11 +1214,14 @@ public void trigger(final ComponentNode component) {
             LOG.info("Performed initial validation of all components in {} milliseconds", millis);
 
+            // Trigger flow analysis to occur every 5 minutes.
+            flowAnalysisThreadPool.scheduleWithFixedDelay(new TriggerFlowAnalysisTask(flowManager.getFlowAnalyzer(), rootProcessGroupSupplier), 5, 5, TimeUnit.MINUTES);
```
With the potential impact on performance, this interval should be promoted to a configurable value in nifi.properties.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/flow/StandardFlowManager.java:
```diff
@@ -519,6 +524,74 @@ public ReportingTaskNode createReportingTask(final String type, final String id,
+            String classloaderIsolationKey
+    ) {
+        if (type == null || id == null || bundleCoordinate == null) {
+            throw new NullPointerException();
```
An empty `NullPointerException` usually  connotes an error that should be corrected with better code handling. Recommend using `IllegalArgumentException` with a message, or `Objects.requireNonNull()` with a message.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/analyzeflow/AbstractFlowAnalysisIT.java:
```diff
@@ -0,0 +1,127 @@
+import java.util.Collections;
+import java.util.UUID;
+
+public abstract class AbstractFlowAnalysisIT extends FrameworkIntegrationTest {
```
Instead of introducing these framework integration tests, recommend creating tests in the `nifi-system-test-suite` module for automated execution. These integration tests are not run automatically and should be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4597,6 +4611,246 @@ public Response deleteReplaceProcessGroupRequest(
+                        serviceFacade.analyzeProcessGroup(analyzedGroupId);
+                        asyncRequest.markStepComplete();
+                    } catch (final Exception e) {
+                        logger.error("Failed to run flow analysis on process group " + processGroupId, e);
```
```suggestion
                        logger.error("Failed to run flow analysis on process group {}", processGroupId, e);
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4597,6 +4611,246 @@ public Response deleteReplaceProcessGroupRequest(
+    @GET
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("flow-analysis/{requestId}")
```
This path overlaps with the `POST` request for Process Groups, perhaps changing the path to `flow-analysis-requests/{requestId}` would be better.
```suggestion
    @Path("{id}/flow-analysis-requests/{requestId}")
```

tpalfy replied:
The overlap is kind of intentional. These 3 methods are really the CURD (without U) of the flow analysis requests.
But I see the point that the POST needs a group id, the GET and the DELETE needs a request id.

So maybe the paths should be these instead:
```java
    @POST
    @Path("{id}/flow-analysis")

    @GET
    @Path("flow-analysis/{requestId}")

    @DELETE
    @Path("flow-analysis/{requestId}")
```

exceptionfactory replied:
That's a good point, but I believe the reason for including the Process Group ID in the path is for authorization handling, @mcgilman should be able to provide further clarification.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4597,6 +4611,246 @@ public Response deleteReplaceProcessGroupRequest(
+    @POST
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("flow-analysis/{processGroupId}")
```
In the context of this Process Group Resource, the path should include the ID first, followed by `flow-analysis-requests`:
```suggestion
    @Path("{id}/flow-analysis-requests")
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4597,6 +4611,246 @@ public Response deleteReplaceProcessGroupRequest(
+    @DELETE
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("flow-analysis/{requestId}")
```
```suggestion
    @Path("{id}/flow-analysis-requests/{requestId}")
```

exceptionfactory reviewed (commented):
Thanks for putting together this new feature @tpalfy! There is a lot here, so I provide some initial feedback on the `nifi-api` and Framework implementation structure. Most of it looks straightforward, but I highlighted a couple questions about custom UI extension-related properties, and some REST API paths. I also highlighted a general concern on package naming, which would impact a number of components.

One other general note, I noticed several integration tests in framework modules. Since those tests are not run automatically, it would be better to create tests in in `nifi-system-test-suite`, which will be run on a regular basis. The Rule implementations in that module look like a useful starting point for system tests.

tpalfy commented on nifi-api/src/main/java/org/apache/nifi/flowanalysis/ComponentAnalysisResult.java:
```diff
@@ -0,0 +1,51 @@
+     * @param message A violation message
+     * @return a new result instance
+     */
+    public static ComponentAnalysisResult newResult(String issueId, String message) {
```
I can change them to simple constructors. A builder feels overcomplicated in this case, especially when there is only one non-mandatory parameter and only 2 different ways of creating a proper instance.

exceptionfactory replied:
Thanks, I agree, switching constructors seems like the best solution given the limited number of arguments for this class.

tpalfy commented on nifi-api/src/main/java/org/apache/nifi/flow/VersionedFlowAnalysisRule.java:
```diff
@@ -0,0 +1,60 @@
+
+public class VersionedFlowAnalysisRule extends VersionedConfigurableExtension {
+
+    private String annotationData;
```
I wasn't sure to be honest. @mcgilman can you confirm that this is only used for custom UI?

exceptionfactory commented on nifi-api/src/main/java/org/apache/nifi/flowanalysis/FlowAnalysisRuleContext.java:
```diff
@@ -0,0 +1,73 @@
+     */
+    boolean isClustered();
+
+    /**
+     * @return the currently configured maximum number of threads that can be
+     * used for executing processors at any given time.
+     */
+    int getMaxTimerDrivenThreadCount();
```
This is a system-level consideration versus a flow-level consideration, so I think we need to reconsider whether it should be part of the API. Although check system-level resource settings might be useful, it is a different set of concerns than flow-level analysis.

exceptionfactory replied:
On further consideration, having the scope of potential evaluation include anything in the application flow configuration flow.json.gz seems reasonable. However, given that this property and others exist as Controller Settings, it seems best to introduce an interface that will contain this property, and potentially other properties. Perhaps something like `ControllerSettingsContext` with a `getControllerSettingsContext()` as a method on `FlowAnalysisRuleContext`?

tpalfy commented on nifi-api/src/main/java/org/apache/nifi/flowanalysis/GroupAnalysisResult.java:
```diff
@@ -0,0 +1,124 @@
+     * @param message A violation message
+     * @return a new analysis result instance tied to the currently analyzed process group
+     */
+    public static GroupAnalysisResult newResultForGroup(String issueId, String message) {
```
Here I will add a Builder. There are multiple but limited and specific ways of creating a proper instance.

tpalfy commented on nifi-framework-api/src/main/java/org/apache/nifi/flowanalysis/AnalyzeFlowState.java:
```diff
@@ -0,0 +1,66 @@
+    CANCELED("Canceled by user", true),
+    ANALYZING("Analyzing flow", false),
+    WAITING("Waiting for flow analyzer to become available", false),
+    COMPLETE("Completed successfully", true);
```
The `AnalyzeFlowRequest` and `AnalyzeFlowState` are actually not needed. They are remnants of a previous version where I didn't use a RequestManager. I'll remove them.

tpalfy commented on nifi-framework-api/src/main/java/org/apache/nifi/flowanalysis/AnalyzeFlowState.java:
```diff
@@ -0,0 +1,66 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.flowanalysis;
```
Turns out no classes in this package are needed any longer. I'll remove the entire package.

tpalfy commented on nifi-framework-api/src/main/java/org/apache/nifi/web/UiExtensionType.java:
```diff
@@ -29,6 +29,7 @@ public enum UiExtensionType {
     ProcessorConfiguration,
     ControllerServiceConfiguration,
     ReportingTaskConfiguration,
+    FlowAnalysisRuleConfiguration,
```
I'll remove it.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/dto/FlowAnalysisRuleDTO.java:
```diff
@@ -0,0 +1,324 @@
+
+@XmlType(name = "flowAnalysisRule")
+public class FlowAnalysisRuleDTO extends ComponentDTO {
+    public static final String VALID = "VALID";
+    public static final String INVALID = "INVALID";
+    public static final String VALIDATING = "VALIDATING";
```
All DTOs are like this and I didn't want to break rank. There is an enum already: `org.apache.nifi.components.validation.ValidationStatus` but it's unavailable here. I guess we didn't want to create another enum.

exceptionfactory replied:
I can understand the existing pattern, although this does seem like a good opportunity introduce an enum.

tpalfy commented on nifi-api/src/main/java/org/apache/nifi/flowanalysis/EnforcementPolicy.java:
```diff
@@ -0,0 +1,29 @@
+ */
+package org.apache.nifi.flowanalysis;
+
+public enum EnforcementPolicy {
```
This is part of the model, it's referenced in `VersionedFlowAnalysisRule`. Maybe I should move it under the `org.apache.nifi.flow` package?

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/flowanalysis/AbstractFlowAnalysisRuleNode.java:
```diff
@@ -0,0 +1,369 @@
+
+            this.state = newState;
+
+            LOG.debug("Successfully {} {}", newState.toString().toLowerCase(), this);
```
I followed the pattern seen in `StandardControllerServiceNode` for this (a flow analysis rule is more similar to controller services in this regard):
```java
public class StandardControllerServiceNode//...
//...
    private void invokeDisable(ConfigurationContext configContext) {
//...
            LOG.debug("Successfully disabled {}", this);
//...
    public CompletableFuture<Void> enable(final ScheduledExecutorService scheduler, final long administrativeYieldMillis) {
                            LOG.info("Successfully enabled {}", service);
...
```

exceptionfactory replied:
Thanks for the reference, it is a minor detail.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/flowanalysis/AbstractFlowAnalysisRuleNode.java:
```diff
@@ -0,0 +1,369 @@
+
+            final ComponentLog componentLog = new SimpleProcessLogger(getIdentifier(), getFlowAnalysisRule());
+
+            componentLog.error("Failed to invoke {} method due to {}", annotation.getSimpleName(), cause);
```
I followed the pattern in `StandardControllerServiceNode` here as well. I checked the UI to make sure the "due to" part is not lost and the message is reported with the exception on the bulletin as well. It is the case so the "due to" part can be omitted.
I'd not change the pattern otherwise though. The expression `"Failed to invoke method [{}]"` would be confusing as the placeholder doesn't receive the method itself but the annotation of the method.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -974,6 +993,9 @@ private ValidationResult validateControllerServiceApi(final PropertyDescriptor d
         return null;
     }
 
+    protected void analyze() {
```
We should avoid method names that are ambiguous. It's very unclear here what sort of analysis should be performed.

tpalfy replied:
That's true but I'm struggling to come up with a good name that's not too long.

This type of analysis conceptually called `FlowAnalysis`. But something like `perfromFlowAnalysis` would suggest that this component wants to do a full-scale analysis. `performFlowAnalysisOnThisComponent` feels too long.
Maybe `performFlowAnalysisOnThis`?

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -776,6 +779,23 @@ protected Collection<ValidationResult> computeValidationErrors(final ValidationC
+                            .explanation(ruleViolation.getViolationMessage())
+                            .build()
+                    )
+                ));
```
We should avoid big, nested blocks of functional code. This would be much cleaner and easier to maintain in an imperative form.

tpalfy replied:
I can change it.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,885 @@
+            }
+
+            if (canRead && canWrite && dataContext.component.persistsState === true) {
+                markup += '<div title="View State" class="pointer view-state-flow-analysis-rule fa fa-tasks"></div>';
```
Do Rules persist State?

tpalfy replied:
Yes the plan is to allow writing rules that need state.

mcgilman replied:
Is there an example of a Rule that is `Stateful`? It would be great to verify the View/Clear State actions work.

tpalfy replied:
Not with this change. Only one rule is added, the `DisallowComponentType` which doesn't hold state.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-flow-analysis-rule.js:
```diff
@@ -0,0 +1,788 @@
+                    .propertytable('setSupportsSensitiveDynamicProperties', flowAnalysisRule.supportsSensitiveDynamicProperties)
+                    .propertytable('loadProperties', flowAnalysisRule.properties, flowAnalysisRule.descriptors, flowAnalysisRuleHistory.propertyHistory)
+                    .propertytable('setPropertyVerificationCallback', function (proposedProperties) {
+                        nfVerify.verify(flowAnalysisRule['id'], flowAnalysisRuleEntity['uri'], proposedProperties, referencedAttributes, handleVerificationResults, $('#flow-analysis-rule-properties-verification-results-listing'));
```
Do Rules support Verification?

mcgilman reviewed (commented):
Thanks for the PR @tpalfy! I just did a quick pass through the front end changes and asked a few questions. Based on the answers, it's likely we can trim down some of the changes here.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,885 @@
+            markup += '<div title="Usage" class="pointer flow-analysis-rule-usage fa fa-book"></div>';
+
+            var hasErrors = !nfCommon.isEmpty(dataContext.component.validationErrors);
+            var hasBulletins = !nfCommon.isEmpty(dataContext.bulletins);
```
The don't.

mcgilman replied:
I think we can safely remove them from consideration in this table.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,885 @@
+
+            // allow policy configuration conditionally
+            if (nfCanvasUtils.isManagedAuthorizer() && nfCommon.canAccessTenants()) {
+                markup += '<div title="Access Policies" class="pointer edit-access-policies fa fa-key"></div>';
```
No we don't.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,885 @@
+            }
+
+            if (isDisabled && canRead && canWrite && dataContext.component.multipleVersionsAvailable === true) {
+                markup += '<div title="Change Version" class="pointer change-version-flow-analysis-rule fa fa-exchange"></div>';
```
In short, yes. I think they shouldn't behave differently in that regard than other extension points.
That being said, I admit I didn't consider this concern before. So if it requires significantly more additional development work then it probably should not be part of this change.

tpalfy replied:
I tested version change - after my latest commit, it works.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,885 @@
+            var markup = '<div title="View Details" class="pointer view-flow-analysis-rule fa fa-info-circle"></div>';
+
+            // always include a button to view the usage
+            markup += '<div title="Usage" class="pointer flow-analysis-rule-usage fa fa-book"></div>';
```
I'm not sure if I understand this question correctly.
Rules should have a "View usage" page like processors and others.
Is this about a different type of documentation? Like custom doc page? I don't plan for them to support custom doc pages.

mcgilman replied:
The `View Usage` action opens the generated documentation for each extension type. If we are going to generate documentation for them, then we should keep this action. If we aren't, or we're not as part of this commit, we should remove this action.

tpalfy replied:
I think `View Usage` should be definitely supported. It turns out that I missed quite a few changes that needs to be done for this to work for flow analysis rules. I added those changes with my latest commit.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/registry/flow/mapping/NiFiRegistryFlowMapper.java:
```diff
@@ -567,7 +589,7 @@ private Map<String, String> mapProperties(final ComponentNode component, final C
     }
 
-    private String encrypt(final String value) {
+    protected String encrypt(final String value) {
```
Is there a reason for changing the visibility on this method?

tpalfy replied:
I needed to override it in 'FlowAnalysisUtil.createMapper'.
Not trivial why this is necessary. I need an encryptor so that encrypted values cannot be viewed as plain text in a rule. But if I used the original `encrypt` method, it would wrap it in and `enc{}` expression which later causes issues.
With this setup every encrypted value is returned as literal of `"*****"`. Although now that I look at the implementation it may have a bug. But that would be the idea.

exceptionfactory replied:
Thanks for the background, that is understandable. It might be worth add a Java Doc comment summarizing the current method implementation, and also noting the different in the overridden method. Just a sentence highlighting the different in return values, along the lines of your comment, would be helpful.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/validation/StandardRuleViolationsManager.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
```
Given the complexity of some operations in this class, it would be helpful to add some basic method comments describe the basic implementation approach.

tpalfy replied:
The interface it implements (`RuleViolationsManager`) does have Javadoc. You mean we should have implementation-specific comments/documentation?

exceptionfactory replied:
Correct, just a sentence or two describing the general memory-backed approach and structure would be helpful for future reference.

exceptionfactory commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-rules/src/main/java/org/apache/nifi/flowanalysis/rules/DisallowComponentType.java:
```diff
@@ -0,0 +1,84 @@
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
```
The `_` character should be avoided as the first character in variable names. Recommend just `descriptors`, or a wrapped declaration like `Collections.unmodifiableList(Arrays.asList(...))`

exceptionfactory replied:
Based on the current Checkstyle rules, there are requirements for packages and methods, but not variables, so it is more of a recommended convention, and looks like something we should improve in the Checkstyle configuration. For now, I recommend avoiding it, but we can address it generally a separate style-focused issue for the repository. Thanks.

exceptionfactory commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-rules/src/main/java/org/apache/nifi/flowanalysis/rules/DisallowComponentType.java:
```diff
@@ -0,0 +1,84 @@
+    public static final PropertyDescriptor COMPONENT_TYPE = new PropertyDescriptor.Builder()
+            .name("component-type")
+            .displayName("Component Type")
+            .description("Components of the given type will produce a rule violation (i.e. they shouldn't exist).")
```
The description should indicate that the Component Type can be a full class name, or a short class name. Matching both may be too far-reaching, as it seems like it be better to just match the full class name, avoiding the possibility and matching components with the same name, but different packages. This could occur if there is a custom subclass that should be preferred.

exceptionfactory reviewed (commented):
Thanks for the latest updates @tpalfy. I reviewed most of the internal framework components and noted a couple additional comments and questions. I have not yet reviewed the latest changes to the REST Resources or UI components, but it looks like the framework components are closer to completion.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-settings.js:
```diff
@@ -1925,6 +1931,872 @@
+                }
+            }
+
+            var bulletinIcon = $(this).find('div.has-bulletins');
```
Thanks for the update to remove the bulletins from the Flow Analysis Rules, however, there are still some remnants here.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/WEB-INF/partials/canvas/settings-content.jsp:
```diff
@@ -23,7 +23,7 @@
         </div>
         <div id="settings-tabs-content">
-            <button id="new-service-or-task" class="add-button fa fa-plus" title="Create a new reporting task controller service" style="display: block;"></button>
+            <button id="new-service-or-task" class="add-button fa fa-plus" title="Create a new reporting task/flow analysis rule controller service" style="display: block;"></button>
```
While not introduced here, I wonder if we should update the title here. Thoughts on just changing this to `Create New`?

tpalfy replied:
I checked and the title of that button is overridden for each controller-level component.
If not it's supposed to be a bug.

I think we could remove the title attribute here.
Or maybe add something that would alert us when it hasn't been overridden. 
Or I can simply revert this change.

mcgilman reviewed (commented):
Thanks for the updates @tpalfy. Just left a couple of minor comments. I should be a +1 on the front-end changes after this.

sardell commented:
> I think we need to address this before merging this PR. I think we could use either of two approaches: 1) Hide the 'Flow Analysis Rules' tab until we have the full feature built into the UI. Or 2) Remove 'Warn' from the available "Enforcement Policy" drop-down, leaving only "Enforce" (or potentially even removing the drop-down for now and hard-coding to "Enforce").

@markap14 I think option 2 with a hard-coded dropdown is a good, easy approach given how far along I am with the [follow-on UI work](https://issues.apache.org/jira/browse/NIFI-11520). I've been connecting with @mcgilman regarding the work I've done so far and he's given me some very helpful initial feedback. I should be ready to share a branch and open a PR very soon after this lands.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-controller-service.js:
```diff
@@ -1079,7 +1118,7 @@
             $.each(referencingComponents, function (_, referencingComponentEntity) {
                 var referencingComponent = referencingComponentEntity.component;
-                if (referencingComponent.referenceType === 'Processor' || referencingComponent.referenceType === 'ReportingTask' || referencingComponent.referenceType === 'ParameterProvider' || referencingComponent.referenceType === 'FlowRegistryClient') {
+                if (referencingComponent.referenceType === 'Processor' || referencingComponent.referenceType === 'ReportingTask' || referencingComponent.referenceType === 'FlowAnalysisRule' || referencingComponent.referenceType === 'ParameterProvider' || referencingComponent.referenceType === 'FlowRegistryClient') {
```
A Flow Analysis Rule is not a schedulable component and it doesn't produce bulletins. I don't think this condition is needed here.

tpalfy replied:
Is this only about bulletins? I thought this also has something to do with the lifecycle of components. All other components available from Controller level are handled here.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-controller-service.js:
```diff
@@ -865,7 +904,7 @@
                 }
             } else {
-                if (referencingComponent.referenceType === 'Processor' || referencingComponent.referenceType === 'ReportingTask') {
+                if (referencingComponent.referenceType === 'Processor' || referencingComponent.referenceType === 'ReportingTask' || referencingComponent.referenceType === 'FlowAnalysisRule') {
```
If an Enabled Flow Analysis Rule references a Controller Service and that service is Disabled, does the Flow Analysis Rule also need to be Disabled?

If not, this is unneeded.

If yes, this will attempt to include it in subsequent calls to update a Controller Service Referencing Components. This happens when a user Disables a Controller Service or when a user Enabled a Controller Service and selects Service and Referencing Components. It would also mean that the back end would need to consider Flow Analysis Rules in this API. I don't think it does currently.

mcgilman replied:
Looking a little more closely here, I think we'd need to make some minor changes to ensure the Flow Analysis Rules are not passed to the back end with the other schedulable components. Instead, it may make sense to include them with the referencing Services but we'd probably want to look at it a little more closely.

tpalfy replied:
Flow Analysis Rules are supposed to work similar to other Controller Service referencing components. If the Controller Service is disabled the Rule should be disabled as well.

Both frontend and backend can handle this with the current implementation.

mcgilman reviewed (commented):
Thanks for the latest round of updates @tpalfy! I've noted a few items below that we should fix prior to merging. I'm also noting here a number front end capabilities that I've highlighted in earlier rounds of review that I've clarified are in fact requirements but are impossible to test currently with the sole Flow Analysis Rule available now does not support them. Those include:

- Referencing a Controller Service from a Property.
- Property verification of a Flow Analysis Rule.
- View/Reset State of a Flow Analysis Rule.
- Custom UI for a Flow Analysis Rule.
- Change Bundle/Version of a Flow Analysis Rule.
- User Specified Sensitive Dynamic Properties

Separately, I see that we've opted to hide `Warn` as an option in the Enforcement Policy. However, that value appears to be the default value for newly created Rules. With the `Warn` option removed, any subsequent changes to the Rule will update this but it looks like the initial value is not what we want it to be.

tpalfy commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-controller-service.js:
```diff
@@ -633,6 +644,33 @@
+                        flowAnalysisRuleGrid.scrollRowIntoView(row);
+
+                        // select the flow analysis rule tab
+                        $('#settings-tabs').find('li:nth-child(4)').click();
```
Fixed.

tpalfy commented:
Hi @mcgilman ,
Thanks for the suggestions.
I added `ControllerServiceReferencingFlowAnalysisRule`, `EnsureFlowAnalysisRuleConfigurationCorrect` and `SensitiveDynamicPropertiesFlowAnalysisRule` in the `nifi-system-test-extensions`.

Building the `nifi-system-test-extensions-bundle` and adding the `nifi-system-test-extensions-nar-2.0.0-SNAPSHOT.nar` to the nifi installation they allow us to test the handling of controller serfices referenced by rules, the verification of rules and the handling of sensitive dynamic properties of rules, respectively.

tpalfy commented:
Hi @mcgilman,

With my last change I tried to addressed all concerns:

> the sole Flow Analysis Rule available now does not support them. Those include:
> 
> - Referencing a Controller Service from a Property.

Added ControllerServiceReferencingFlowAnalysisRule

> - Property verification of a Flow Analysis Rule.

Added EnsureFlowAnalysisRuleConfigurationCorrect

> - View/Reset State of a Flow Analysis Rule.

Added StatefulFlowAnalysisRule

> - Custom UI for a Flow Analysis Rule.

Custom UI is not supported for Flow Analysis Rules

> - Change Bundle/Version of a Flow Analysis Rule.

This can be tested like with other components. We can build a nar (either nifi-standard-nar or nifi-system-test-extensions-nar) with a different version and add to an existing nifi installation.

> - User Specified Sensitive Dynamic Properties

Added SensitiveDynamicPropertiesFlowAnalysisRule

> 
> Separately, I see that we've opted to hide `Warn` as an option in the Enforcement Policy. However, that value appears to be the default value for newly created Rules. With the `Warn` option removed, any subsequent changes to the Rule will update this but it looks like the initial value is not what we want it to be.this but it looks like the initial value is not what we want it to be.

When Flow Analysis Rules are created via the UI the Enforcement Policy will be set to 'Enforce' so the initial value will be that. I understand that this relies on the UI but this is meant to be temporary and I didn't want to overcomplicate it so that the reversal can be done as smoothly as possible.

Please let me know if I have missed something.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:
```diff
@@ -3039,6 +3098,86 @@ public Response getTemplates() {
+    @GET
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("flow-analysis/result")
```
It seems like this path should be pluralized to `/results`:
```suggestion
    @Path("flow-analysis/results")
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:
```diff
@@ -3039,6 +3098,86 @@ public Response getTemplates() {
+    @GET
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("flow-analysis/result/{processGroupId}")
```
```suggestion
    @Path("flow-analysis/results/{processGroupId}")
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:
```diff
@@ -3039,6 +3098,86 @@ public Response getTemplates() {
+        value = "Returns flow analysis results produced by the analysis of a given process group",
+        response = FlowAnalysisResultEntity.class,
+        authorizations = {
+            @Authorization(value = "Read - /controller")
```
```suggestion
            @Authorization(value = "Read - /flow")
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:
```diff
@@ -3039,6 +3098,86 @@ public Response getTemplates() {
+        value = "Returns all flow analysis results currently in effect",
+        response = FlowAnalysisResultEntity.class,
+        authorizations = {
+            @Authorization(value = "Read - /controller")
```
```suggestion
            @Authorization(value = "Read - /flow")
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/impl/StandardFlowAnalysisRuleDAO.java:
```diff
@@ -0,0 +1,370 @@
+                }
+            } catch (IllegalArgumentException iae) {
+                throw new IllegalArgumentException(String.format(
+                        "The specified flow analysis rule state (%s) is not valid. Valid options are 'RUNNING', 'STOPPED', and 'DISABLED'.",
```
```suggestion
                        "The specified flow analysis rule state (%s) is not valid. Valid options are 'ENABLED' or 'DISABLED'.",
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/dto/FlowAnalysisRuleDTO.java:
```diff
@@ -0,0 +1,324 @@
+    private Map<String, PropertyDescriptorDTO> descriptors;
+    private Set<String> sensitiveDynamicPropertyNames;
+
+    private String customUiUrl;
+    private String annotationData;
```
These two properties can be removed.

exceptionfactory reviewed (commented):
Thanks for the additional feedback @mcgilman, and thanks for the updates @tpalfy. Those changes appear to address the issues noted.

Running through one more review, I noted a few more properties that need to be removed, and a couple annotated documentation adjustments. All of these should be minor, and following that, this looks ready to be merged.

exceptionfactory reviewed (approved):
Thanks for making the last round of updates and persevering through multiple rounds of feedback @tpalfy! The latest version sufficiently addresses the current feedback, and provides a good basis for subsequent user interface implementation.

Thanks @markap14, @mcgilman, and @sardell for the reviews and feedback along the way!

+1 merging
